clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -99,8 +99,8 @@
     private final ConcurrentLinkedQueue<CompletedFetch> completedFetches;
     private final Deserializer<K> keyDeserializer;
     private final Deserializer<V> valueDeserializer;
-
     private PartitionRecords nextInLineRecords = null;
+    private ExceptionMetadata nextInLineExceptionMetadata = null;
```

Comment:
I am wondering that would it be simpler if we just peek the first `CompletedFetch` in the `completedFetches`, and only remove it after it is parsed. The logic would be:
1. If `parseCompleted(completedFetches.peek())` did not throw exception, remove the CompletedFetch.
2. If `parseCompleted(completedFetches.peek())` threw exception and 'fetched.isEmpty() == false', catch the exception, return the ConsumerRecords that has already been fetched.
3. If `parseCompleted(completedFetches.peek())` threw exception and 'fetched.isEmpty() == true', catch the exception, remove the first entry in `completedFetches` and re-throw the exception.

This way we don't need to cache any result.